llvm/docs/GitHub.rst

```diff
@@ -105,7 +105,9 @@ branch that's the root of the Pull Request during the review. This action will
 make the context of the old changes and comments harder to find and read.
 
 Sometimes, a rebase might be needed to update your branch with a fix for a test
-or in some dependent code.
+or in some dependent code. This is especially encouraged if it turns out that
+the upstream base commit used for your branch had test failures, meaning the
+pre-commit CI results are not useful.
```

Comment:
It may be phrased a bit strongly: I often see test failures in pre-commit CI in other projects like  while modifying MLIR: as long as MLIR and Flang tests are passing, the pre-commit CI results aren't "useless".
I would say only when there is a build failure and the tests can't run that the results aren't useful anymore.

That said the advice isn't bad for beginner.

Reply:
How about "not useful" => "misleading"? The big issue with failures is that you see that red cross and have to click through for more details to see if it's a "real" problem (an issue with the PR author's patch) or not.

Reply:
What about:

> This is especially encouraged if it turns out that the upstream base commit used for your branch had test failures, meaning the pre-commit CI results can't be shown as cleanly passing.

Reply:
Why are we encouraging rebase versus merging main into your branch to address this issue? IMO, we should explicitly discourage any use of rebase.

Reply:
@joker-eph: Thanks, I've adopted that change.

@jyknight: Given none of our current documentation discusses introducing merge commits (even if they disappear before hitting the main branch) I think this would be more complex to explain. For me, I'm only really used to this rebase-focused workflow so I'm not actually sure how GitHub would handle your suggestion in its UI - both when showing changes for review and in terms of allowing "squash and merge" to commit. I'd love to learn more though. Is there a particular advantage to doing it as you suggest?

Reply:
Github provides very poor support for tracing the history of a PR, and looking at changes since you last reviewed, across a series of rebases. This is very much unlike gerrit/phab, where that's the primary way they provide to see changes.

But it provides good support for looking at a series of new patches pushed on top.

As a reviewer, I want two things:
1. look at new changes that were made in response to comments
2. look at the complete diff from mainline.

The first is easier when rebasing is avoided, and the latter is not harmed by the presence of merge-commits in the PR history. You can see an example of a PR with merge-commits in my pending PR here: https://github.com/llvm/llvm-project/pull/74275. (There's plenty more examples, though.)

Here's an example PR with lots of history and multiple rebases: https://github.com/llvm/llvm-project/pull/65534. I find it very hard to look at that and see what's actually changed. Note that all the commits in the timeline are listed as "3 weeks ago". But they're actually older. There are "force-pushed" notices interspersed elsewhere, but it's still quite difficult to reconstruct when they were actually introduced.

Also, once you have a big patch with lots of changes like that, doing rebases starts getting really noisy in notifications -- it resends the list of every commit in the PR, since they've all been pushed fresh.

Regarding "squash-and-merge", that takes the complete diff from mainline, and makes a new commit with that -- so it works just fine if you have merges in your PR branch.

Reply:
Chiming in as the author of aforementioned https://github.com/llvm/llvm-project/pull/65534: I've decided to go with `git merge` instead of `git rebase` from now on exactly due to the UI reason -- i too find it much easier to navigate through changes this way, when imo rebasing just breaks history.